src/lib.rs

```diff
@@ -1,134 +1,275 @@
-                        let ident = &v.ident;
-                        match &v.fields {
-                            syn::Fields::Named(f) => {
-                                let fields: Vec<_> = f
-                                    .named.clone()
-                                    .into_iter()
-                                    .map(|f| f.ident.expect("Every field should have a name"))
-                                    .collect();
+                    .map(|variant| {
+                        let variant_ident = &variant.ident;
+                        let name = quote! { Self::#variant_ident };
+
+                        match &variant.fields {
+                            Fields::Named(fields) => {
+                                let body = self.generate_struct(name, fields);
+                                let fields_destructure = Self::prepare_field_named(fields);
+
                                 quote! {
-                                    Self::#ident{#(#fields),*} => Self::#ident{#(#fields: #fields.clone()),*}
+                                    Self::#variant_ident { #(#fields_destructure),* } => { #body }
                                 }
                             }
-                            syn::Fields::Unnamed(f) => {
-                                let fields: Vec<_> = (0..f.unnamed.len())
-                                    .into_iter()
-                                    .map(|n| format_ident!("field{}", n))
-                                    .collect();
+                            Fields::Unnamed(fields) => {
+                                let body = self.generate_tuple(name, fields);
+                                let fields_temp = Self::prepare_field_unnamed(fields);
+
                                 quote! {
-                                    Self::#ident(#(#fields),*) => Self::#ident(#(#fields.clone()),*)
+                                    Self::#variant_ident ( #(#fields_temp),* ) => { #body }
                                 }
                             }
-                            syn::Fields::Unit => quote! {Self::#ident => Self::#ident},
+                            Fields::Unit => {
+                                let body = self.generate_unit(name);
+                                quote! { Self::#variant_ident => { #body } }
+                            }
                         }
                     })
                     .collect();
+
                 quote! {
-                    fn clone(&self) -> Self {
-                        match self {
-                            #(#body),*
-                        }
+                    match self {
+                        #(#bodies),*
                     }
                 }
             }
+            Data::Union(_) => todo!("Unions are not supported"),
+        };
+
+        self.generate_signature(body)
+    }
+
+    fn prepare_field_named(fields: &FieldsNamed) -> impl Iterator<Item = TokenStream> + '_ {
```

Comment:
These could move to a Trait implemented for Fields or FieldsNamed, but maybe it does make more sense to keep them here, as they would be the only two we can do that for.

Reply:
I re-worked these parts largely.

Reply:
These functions don't exist anymore, so I will let you re-review this.

Reply:
Yeah, I think it is reasonable to keep them there for now, if we know for sure what we want from them, we could move them to a separate module.